In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

In [2]:
df = pd.read_excel('E Commerce Dataset.xlsx')
df.head()

CustomerID  Tenure PreferredLoginDevice  CityTier  WarehouseToHome  \
0       50001     4.0         Mobile Phone         3              6.0   
1       50002     NaN                Phone         1              8.0   
2       50003     NaN                Phone         1             30.0   
3       50004     0.0                Phone         3             15.0   
4       50005     0.0                Phone         1             12.0   

  PreferredPaymentMode  Gender  HourSpendOnApp  NumberOfDeviceRegistered  \
0           Debit Card  Female             3.0                         3   
1                  UPI    Male             3.0                         4   
2           Debit Card    Male             2.0                         4   
3           Debit Card    Male             2.0                         4   
4                   CC    Male             NaN                         3   

     PreferedOrderCat  SatisfactionScore MaritalStatus  NumberOfAddress  \
0  Laptop & Accessory                  2        Single                9   
1              Mobile                  3        Single                7   
2              Mobile                  3        Single                6   
3  Laptop & Accessory                  5        Single                8   
4              Mobile                  5        Single                3   

   Complain  OrderAmountHikeFromlastYear  CouponUsed  OrderCount  \
0         1                         11.0         1.0         1.0   
1         1                         15.0         0.0         1.0   
2         1                         14.0         0.0         1.0   
3         0                         23.0         0.0         1.0   
4         0                         11.0         1.0         1.0   

   DaySinceLastOrder  CashbackAmount  Churn  
0                5.0          159.93      1  
1                0.0          120.90      1  
2                3.0          120.28      1  
3                3.0          134.07      1  
4                3.0          129.60      1

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5630 entries, 0 to 5629
Data columns (total 20 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   CustomerID                   5630 non-null   int64  
 1   Tenure                       5366 non-null   float64
 2   PreferredLoginDevice         5630 non-null   object 
 3   CityTier                     5630 non-null   int64  
 4   WarehouseToHome              5379 non-null   float64
 5   PreferredPaymentMode         5630 non-null   object 
 6   Gender                       5630 non-null   object 
 7   HourSpendOnApp               5375 non-null   float64
 8   NumberOfDeviceRegistered     5630 non-null   int64  
 9   PreferedOrderCat             5630 non-null   object 
 10  SatisfactionScore            5630 non-null   int64  
 11  MaritalStatus                5630 non-null   object 
 12  NumberOfAddress              5630 non-null   int64  
 13  Complain          

In [7]:
df.nunique()

CustomerID                     5630
Tenure                           36
PreferredLoginDevice              3
CityTier                          3
WarehouseToHome                  34
PreferredPaymentMode              7
Gender                            2
HourSpendOnApp                    6
NumberOfDeviceRegistered          6
PreferedOrderCat                  6
SatisfactionScore                 5
MaritalStatus                     3
NumberOfAddress                  15
Complain                          2
OrderAmountHikeFromlastYear      16
CouponUsed                       17
OrderCount                       16
DaySinceLastOrder                22
CashbackAmount                 2586
Churn                             2
dtype: int64

In [9]:
for col in df.drop(['CustomerID', 'CashbackAmount'], axis=1).columns:
  print('Column Name :', col)
  print(df[col].unique())
  print('-------------------------------------------------------------')

Column Name : Tenure
[ 4. nan  0. 13. 11.  9. 19. 20. 14.  8. 18.  5.  2. 30.  1. 23.  3. 29.
  6. 26. 28.  7. 24. 25. 10. 15. 22. 27. 16. 12. 21. 17. 50. 60. 31. 51.
 61.]
-------------------------------------------------------------
Column Name : PreferredLoginDevice
['Mobile Phone' 'Phone' 'Computer']
-------------------------------------------------------------
Column Name : CityTier
[3 1 2]
-------------------------------------------------------------
Column Name : WarehouseToHome
[  6.   8.  30.  15.  12.  22.  11.   9.  31.  18.  13.  20.  29.  28.
  26.  14.  nan  10.  27.  17.  23.  33.  19.  35.  24.  16.  25.  32.
  34.   5.  21. 126.   7.  36. 127.]
-------------------------------------------------------------
Column Name : PreferredPaymentMode
['Debit Card' 'UPI' 'CC' 'Cash on Delivery' 'E wallet' 'COD' 'Credit Card']
-------------------------------------------------------------
Column Name : Gender
['Female' 'Male']
--------------------------------------------------------

# OBSERVATIONS:

1. **PreferredLoginDevice**: The values 'Mobile Phone' and 'Mobile' both likely refer to mobile devices, as it's not possible to log in using a landline. To ensure consistency, standardize these values by replacing both with 'Mobile.'
   
2. **WarehouseToHome**: There are some unusually high values, such as 126 and 127, which are unlikely since most people don’t travel such long distances to purchase goods. These seem to be input errors, so adjust the values: replace 126 with 26 and 127 with 27.
   
3. **PreferredPaymentMode**:
   - 'CC' and 'Credit Card' both refer to the same payment method, so standardize them as 'CC.'
   - 'Cash on Delivery' and 'COD' also indicate the same payment method; replace both with 'COD.'
   - For consistency, rename 'Debit Card' as 'DC.'
   
4. **PreferedOrderCat**:
   - 'Mobile' and 'Mobile Phone' refer to the same category, so change both to 'Mobile.'
   - Simplify 'Laptop & Accessory' to 'Laptop' for convenience.
   
5. **NumberOfAddress**: Some values exceed 10, which seems implausible. These values require further investigation to verify their accuracy.

In [13]:
df['PreferredLoginDevice'].replace({'Mobile Phone':'Mobile', 'Phone':'Mobile'}, inplace=True)
df['WarehouseToHome'].replace({126:26, 127:27}, inplace=True)
df['PreferredPaymentMode'].replace({'Credit Card':'CC', 'Cash on Delivery':'COD', 'Debit Card':'DC' }, inplace=True)
df['PreferedOrderCat'].replace({'Mobile Phone':'Mobile', 'Laptop & Accessory':'Laptop'}, inplace=True)

C:\Users\91932\AppData\Local\Temp\ipykernel_4520\2726075726.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['PreferredLoginDevice'].replace({'Mobile Phone':'Mobile', 'Phone':'Mobile'}, inplace=True)
C:\Users\91932\AppData\Local\Temp\ipykernel_4520\2726075726.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are settin

In [15]:
df[df.NumberOfAddress>3].shape[0]

2612

There seems to be 2612 customers with more than 3 addresses registered in the ecommerce platform, in their name. This seems very unlikely. But we cannot rule out the following possibilities:

The customer might have multiple properties(houses) in their name.
Slight change in the address text.
Addresses of workplace/workplaces if the customer works shifts.
Ordering for someone else
Someone else must be ordering in the customer's name name(either other family member with the customer's knowledge or in the rare case fraud transactions by outsiders).
In the above mention cases, multiple addresses might be added.

If we have some other information like the actual list of addresses registered or the locations where the orders were made, number of family members, number of children, occupation related information, annual income etc., we might get a little bit insight into the possibility of customers have multiple addresses.

However, since we do not have them, we need to explore with the available information and reach a conclusion. We shall analyse more in the EDA process and come to a conclusion.

In [17]:
df.isna().sum()

CustomerID                       0
Tenure                         264
PreferredLoginDevice             0
CityTier                         0
WarehouseToHome                251
PreferredPaymentMode             0
Gender                           0
HourSpendOnApp                 255
NumberOfDeviceRegistered         0
PreferedOrderCat                 0
SatisfactionScore                0
MaritalStatus                    0
NumberOfAddress                  0
Complain                         0
OrderAmountHikeFromlastYear    265
CouponUsed                     256
OrderCount                     258
DaySinceLastOrder              307
CashbackAmount                   0
Churn                            0
dtype: int64

In [19]:
missing_cols = [col for col in df.columns if df[col].isnull().any()]
missing_cols

['Tenure',
 'WarehouseToHome',
 'HourSpendOnApp',
 'OrderAmountHikeFromlastYear',
 'CouponUsed',
 'OrderCount',
 'DaySinceLastOrder']

In [21]:
percent = df[missing_cols].isnull().sum() * 100/len(df)
missing_df = pd.DataFrame({'column':missing_cols,
                           'no. of missing values' : df[missing_cols].isnull().sum().values,
                           'percentage' : percent.values})
missing_df

column  no. of missing values  percentage
0                       Tenure                    264    4.689165
1              WarehouseToHome                    251    4.458259
2               HourSpendOnApp                    255    4.529307
3  OrderAmountHikeFromlastYear                    265    4.706927
4                   CouponUsed                    256    4.547069
5                   OrderCount                    258    4.582593
6            DaySinceLastOrder                    307    5.452931

We can see that



1.**HourSpendOnApp** is normally distributed -> Fill missing values with mean



2.**Tenure and OrderAmountHikeFromlastY**ear are moderately skewed -> Fill missing values with mean



3.**WarehouseToHome, CouponUsed, OrderCount and DaySinceLastOrder** are highly skewed -> Fill missing values with Knn imputer since they are correlateddian

In [33]:
round(df.HourSpendOnApp.mean()), round(df.Tenure.mean()), round(df.OrderAmountHikeFromlastYear.mean())
     

(3, 10, 16)

In [35]:

df.CouponUsed.median(), df.OrderCount.median(), df.DaySinceLastOrder.median(),df.WarehouseToHome.median()

(1.0, 2.0, 3.0, 14.0)

In [39]:
df.fillna({'HourSpendOnApp':round(df.HourSpendOnApp.mean()), 
           'Tenure':round(df.Tenure.mean()), 
           'OrderAmountHikeFromlastYear':round(df.OrderAmountHikeFromlastYear.mean()),
           'WarehouseToHome':df.WarehouseToHome.median(), 
           'CouponUsed':df.CouponUsed.median(), 
           'OrderCount':df.OrderCount.median(), 
           'DaySinceLastOrder':df.DaySinceLastOrder.median()}, inplace=True)

In [41]:
df.isnull().sum()

CustomerID                     0
Tenure                         0
PreferredLoginDevice           0
CityTier                       0
WarehouseToHome                0
PreferredPaymentMode           0
Gender                         0
HourSpendOnApp                 0
NumberOfDeviceRegistered       0
PreferedOrderCat               0
SatisfactionScore              0
MaritalStatus                  0
NumberOfAddress                0
Complain                       0
OrderAmountHikeFromlastYear    0
CouponUsed                     0
OrderCount                     0
DaySinceLastOrder              0
CashbackAmount                 0
Churn                          0
dtype: int64

In [43]:
cleaned_data='cleaned_data.xlsx'
df.to_excel(cleaned_data,index=False)
print(f"Dataset saved successfully to {cleaned_data}")

Dataset saved successfully to cleaned_data.xlsx
